File: scala/sum-of-multiples/src/main/scala/SumOfMultiples.scala
object SumOfMultiples {
  def multiples(nr:Int, given:Int) =
    Stream.from(1).map(_ * nr).takeWhile(_ < given)

  def sumOfMultiples(nrs: Seq[Int], given: Int) =
    nrs.toSet.flatMap(multiples(_:Int, given)).sum
}

File: java/sum-of-multiples/src/main/java/SumOfMultiples.java
import java.util.function.IntFunction;
import java.util.stream.IntStream;

public class SumOfMultiples {

    public static int sum(int limit, int[] xs) {
        return IntStream.of(xs)
                .flatMap(SumOfMultiples.multiples(limit))
                .distinct()
                .sum();
    }

    public static IntFunction<IntStream> multiples(int limit) {
        return x ->
                IntStream.iterate(x, y -> x + y)
                        .limit( limit % x == 0 ? limit/x - 1: limit/x );
    }
}
